DOCS-481: Refactor Vision Service (#883)
viamrobotics/docs: docs/services/vision/detection.md

#### TFLite Model Limitations

We strongly recommend that you package your `.tflite` model with metadata in [the standard form](https://github.com/tensorflow/tflite-support/blob/560bc055c2f11772f803916cb9ca23236a80bf9d/tensorflow_lite_support/metadata/metadata_schema.fbs).

In the absence of metadata, your TFLite model must satisfy the following requirements:

- A single input tensor representing the image of type UInt8 (expecting values from 0 to 255) or Float 32 (values from -1 to 1).
- At least 3 output tensors (the rest won’t be read) containing the bounding boxes, class labels, and confidence scores (in that order).
- Bounding box output tensor must be ordered [x x y y], where x is an x-boundary (xmin or xmax) of the bounding box and the same is true for y.
  Each value should be between 0 and 1, designating the percentage of the image at which the boundary can be found.

These requirements are satisfied by a few publicly available model architectures including EfficientDet, MobileNet, and SSD MobileNet V1.
You can use one of these architectures or build your own.

## Configuration

### Add the service and detector

> 

Reviewer: It seems like detection.md gets this section about configuration, but classification does not. It would be useful if classification also had a section like it - or if it were put at the general level.

Author: Yes - I wanted to run this by you before adding creating all the pages - just in case it wasn't a good idea for some reason

   }
 ]
}
```

Click **SAVE CONFIG** and head to the **COMPONENTS** tab.

{{%expand "You can also configure the entire Vision Service and detector in raw JSON" %}}

To add a vision model to your robot, add the `name`, `type`, and `parameters` of the desired detector to the `register_models` in the attributes field of the Vision Service config.
For example:

``` json {class="line-numbers linkable-line-numbers"}
"services": [
    {
        "name": "vision1",
        "type": "vision",
        "attributes": {
          "register_models": [
            {
              "name": "my_color_detector",
              "type": "color_detector",
              "parameters": {
                "detect_color" : "#A3E2FF",
                "hue_tolerance_pct": 0.06,
                "segment_size_px": 100
              }
            },
            {
              "name": "my_classifier",
              "type": "tflite_classifier",
              "parameters": {
                "model_path" : "/path/to/model.tflite",
                "label_path": "/path/to/labels.txt",
                "num_threads": 1
              }
            }
          ]
        }
    }
]
```

{{% /expand%}}

> ### Add a camera component and a "transform" model

Reviewer: the same can be done for a classification

Any camera that can return 2D images can use 2D object detection.

The service provides different types of detectors, both heuristic and machine-learning based, so that you can create, register, and use detectors for any object you may need to identify.

The returned detections consist of the bounding box around the identified object, as well as its label and confidence score:

- `x_min`, `y_min`, `x_max`, `y_max` (int): specify the bounding box around the object.
- `class_name` (string): specifies the label of the found object.
- `confidence` (float): specifies the confidence of the assigned label between 0.0 and 1.0.

## Detector Types

You can use the following types of detectors:

- [**color_detector**](#color-detector): A heuristic based detector that draws boxes around objects according to their hue (does not detect black, gray, and white).
- [**tflite_detector**](#tflite-detector): A machine-learning based detector that draws bounding boxes according to the specified tensorflow-lite model file available on the robot’s hard drive.

> ### Color detector

Reviewer 2: ```suggestion
### `color_detector`
```

A heuristic based detector that draws boxes around objects according to their hue.

> Color detectors do not detect black, perfect greys (greys where the red, green, and blue color component values are equal), or white.

Author: ohhhh grays? I didn't even write this! I think...

```suggestion
Color detectors do not detect black, perfect grays (grays where the red, green, and blue color component values are equal), or white.
```

It only detects hues found on the color wheel.

{{% alert title="Note" color="note" %}}
Object colors can vary dramatically based on the light source.
We recommend you verify the desired color detection value under actual lighting conditions.
To determine the color value from the actual cam component image, you can use a pixel color tool, like [Color Picker for Chrome](https://chrome.google.com/webstore/detail/color-picker-for-chrome/clldacgmdnnanihiibdgemajcfkmfhia).

If the color is not reliably detected, increase the `hue_tolerance_pct`.
{{< /alert >}}

<br>

These are the available parameters in the detector's configuration. For an example see [Configuration](#configuration).

``` json {class="line-numbers linkable-line-numbers"}
{
    "register_models": [
        {
            "name": "<detector_name>",
            "type": "color_detector",
            "parameters": {
              "detect_color" : "#ABCDEF",
              "hue_tolerance_pct": <number>,
              "segment_size_px": <integer>,
              "saturation_cutoff_pct": <number>,
              "value_cutoff_pct" <number>
            }
        }
    ]
}
```

> | Parameter | Description |

Reviewer 2: Would also suggest adding an inclusion column here

| `segment_size_px` | An integer that sets a minimum size (in pixels) of a contiguous color region to be detected, and filters out all other found objects below that size. |
| `saturation_cutoff_pct (optional)` | A number > 0.0 and <= 1.0 which defines the minimum saturation before a color is ignored. Defaults to 0.2. |
| `value_cutoff_pct (optional)` | A number > 0.0 and <= 1.0 which defines the minimum value before a color is ignored. Defaults to 0.3. |

{{% alert title="Note" color="note" %}}

**hue_tolerance_pct**, **saturation_cutoff_pct**, and **value_cutoff_pct** refer to hue, saturation, and value (brightness) in the HSV Color Model, but do not set color values in Viam.

**hue_tolerance_pct** specifies the exactness of the color match to **detect_color**.

The optional **saturation_cutoff_pct** and **value_cutoff_pct** attributes specify cutoff thresholds levels for saturation and brightness, rather than specifying color saturation and brightness as they do in the standard HSV Color Model.

{{% /alert %}}

> ### TFLite detector

Reviewer 2: ```suggestion
### `tflite_detector`
```

A machine-learning based detector that draws bounding boxes according to the specified tensorflow-lite model file available on the robot’s hard drive.

These are the available parameters in the detector's configuration. For an example see [Configuration](#configuration).

``` json {class="line-numbers linkable-line-numbers"}
{
    "register_models": [
        {
            "name": "<detector_name>",
            "type": "tflite_classifier",
            "parameters": {
              "model_path" : "/path/to/model.tflite",
              "label_path": "/path/to/labels.txt",
              "num_threads": <number>
            }
        }
    ]
}
```

| Parameter | Description |
| --------- | ----------- |
| `model_path`| Required. The path to the `.tflite model` file, as a `string`. |
| `num_threads`| An integer that defines how many CPU threads to use to run inference. The default value is 1. |
| `label_path`| The path to a `.txt` file that holds class labels for your TFLite model, as a `string`. The SDK expects this text file to contain an ordered listing of the class labels. Without this file, classes will read as "1", "2", and so on. |

> #### TFLite Model Limitations

Reviewer 2: ```suggestion
#### `tflite_model` Limitations
```

We strongly recommend that you package your `.tflite` model with metadata in [the standard form](https://github.com/tensorflow/tflite-support/blob/560bc055c2f11772f803916cb9ca23236a80bf9d/tensorflow_lite_support/metadata/metadata_schema.fbs).

> In the absence of metadata, your TFLite model must satisfy the following requirements:

Reviewer 2: ```suggestion
In the absence of metadata, your `.tflite` model must satisfy the following requirements:
```

                "model_path" : "/path/to/model.tflite",
                "label_path": "/path/to/labels.txt",
                "num_threads": 1
              }
            },
            {
                "name": "my_segmenter",
                "type": "radius_clustering_segmenter",
                "parameters": {
                    "min_points_in_plane": 2,
                    "min_points_in_segment": 2,
                    "clustering_radius_mm": 3.0,
                    "mean_k_filtering": 0
                }
            }
          ]
        }
    }
]
```

{{% /expand%}}

### Add a camera component and a "transform" model

You cannot interact directly with the [Vision Service](/services/vision/).
To be able to interact with the Vision Service you must:

1. Configure a physical [camera component](../../../components/camera) to wrap the service.
2. Configure a [transform camera](../../../components/camera/transform) to view output from the detector overlaid on images from the physical camera.

After adding the component and its attributes, click **SAVE CONFIG**.

Wait for the robot to reload, and then go to the **CONTROL** tab to test the stream of detections.

## Code

> The following code gets the robot’s Vision Service and then runs a color detector vision model on an image from the robot's camera named `"camera_1"` in this example.

Reviewer 2: ```suggestion
The following code gets the robot’s Vision Service and then runs a color detector vision model on an image from the robot's camera, named `"camera_1"` in this example.
```
Opt i'm not sure of grammar here

Author: ```suggestion
The following code gets the robot’s Vision Service and then runs a color detector vision model on an image from the robot's camera `"camera_1"`:
```

{{< tabs >}}
{{% tab name="Python" %}}

```python {class="line-numbers linkable-line-numbers"}
from viam.services.vision import VisionServiceClient, VisModelConfig, VisModelType

robot = await connect()
# grab camera from the robot
cam1 = Camera.from_robot(robot, "cam1")
# grab Viam's vision service which has the TFLite detector already registered
vision = VisionServiceClient.from_robot(robot)

print("Vision Resources:")

> print(await vision.get_detector_names())

Reviewer: get_segmenter_names()

---
title: "Detection (or 2D object detection)"
linkTitle: "Detection"
weight: 10
type: "docs"
description: "Select an algorithm that identifies objects in a 2D image and adds bounding boxes around identified objects."
tags: ["vision", "computer vision", "CV", "services", "detection"]
# SMEs: Bijan, Khari
---

_2D Object Detection_ is the process of taking a 2D image from a camera and identifying and drawing a box around the distinct "objects" of interest in the scene.
Any camera that can return 2D images can use 2D object detection.

> The service provides different types of detectors, both heuristic and machine-learning based, so that you can create, register, and use detectors for any object you may need to identify.

Reviewer 3: ```suggestion
The service provides different types of detectors, both heuristic and machine-learning-based, so that you can create, register, and use detectors for any object you may need to identify.
```

Author: ```suggestion
The service provides different types of detectors, both based on heuristics and machine learning, so that you can create, register, and use detectors for any object you may need to identify.
```

The returned detections consist of the bounding box around the identified object, as well as its label and confidence score:

- `x_min`, `y_min`, `x_max`, `y_max` (int): specify the bounding box around the object.
- `class_name` (string): specifies the label of the found object.
- `confidence` (float): specifies the confidence of the assigned label.
  Between `0.0` and `1.0`, inclusive.

## Detector Types

You can use the following types of detectors:

> - [**color_detector**](#color-detector): A heuristic based detector that draws boxes around objects according to their hue (does not detect black, gray, and white).

Author: ```suggestion
- [**color_detector**](#color_detector): A heuristic based detector that draws boxes around objects according to their hue (does not detect black, gray, and white).
```

> - [**tflite_detector**](#tflite-detector): A machine-learning based detector that draws bounding boxes according to the specified .tflite model file available on the robot’s hard drive.

Author: ```suggestion
- [**tflite_detector**](#tflite_detector): A machine-learning based detector that draws bounding boxes according to the specified .tflite model file available on the robot’s hard drive.
```

> - [**tflite_detector**](#tflite_detector): A machine-learning based detector that draws bounding boxes according to the specified .tflite model file available on the robot’s hard drive.

Reviewer 3: ```suggestion
- [**tflite_detector**](#tflite_detector): A machine-learning-based detector that draws bounding boxes according to the specified .tflite model file available on the robot’s hard drive.
```

### `color_detector`

A heuristic based detector that draws boxes around objects according to their hue.
Color detectors do not detect black, perfect grays (grays where the red, green, and blue color component values are equal), or white.
It only detects hues found on the color wheel.

{{% alert title="Note" color="note" %}}
Object colors can vary dramatically based on the light source.
We recommend you verify the desired color detection value under actual lighting conditions.

> To determine the color value from the actual cam component image, you can use a pixel color tool, like [Color Picker for Chrome](https://chrome.google.com/webstore/detail/color-picker-for-chrome/clldacgmdnnanihiibdgemajcfkmfhia).

Reviewer 3: ```suggestion
To determine the color value from the actual camera component image, you can use a pixel color tool, like [Color Picker for Chrome](https://chrome.google.com/webstore/detail/color-picker-for-chrome/clldacgmdnnanihiibdgemajcfkmfhia).
```
[nit]

            {
              "name": "my_color_detector",
              "type": "color_detector",
              "parameters": {
                "detect_color" : "#A3E2FF",
                "hue_tolerance_pct": 0.06,
                "segment_size_px": 100
              }
            },
            {
              "name": "my_classifier",
              "type": "tflite_classifier",
              "parameters": {
                "model_path" : "/path/to/model.tflite",
                "label_path": "/path/to/labels.txt",
                "num_threads": 1
              }
            },
            {
                "name": "my_segmenter",
                "type": "radius_clustering_segmenter",
                "parameters": {
                    "min_points_in_plane": 2,
                    "min_points_in_segment": 2,
                    "clustering_radius_mm": 3.0,
                    "mean_k_filtering": 0
                }
            }
          ]
        }
    }
]
```

{{% /expand%}}

### Add a camera component and a "transform" model

You cannot interact directly with the [Vision Service](/services/vision/).
To be able to interact with the Vision Service you must:

> 1. Configure a physical [camera component](../../../components/camera) to wrap the service.

Reviewer 3: What does it mean for a physical camera to wrap the service? Is that right?

Author: You're right - it should just say configure a physical cam component

```suggestion
1. Configure a physical [camera component](../../../components/camera).
```

2. Configure a [transform camera](../../../components/camera/transform) to view output from the detector overlaid on images from the physical camera.

After adding the component and its attributes, click **SAVE CONFIG**.

Wait for the robot to reload, and then go to the **CONTROL** tab to test the stream of detections.

## Code

> The following code gets the robot’s Vision Service and then runs a color detector vision model on an image from the robot's camera `"camera_1"`:

Reviewer 3: [probably out of PR scope] This usage of "Vision Service" makes me wonder whether we should capitalize services in the context of "the Viam Vision Service allows you to do fancy things" but not in the context of a given instance of the service

Author: Let's discuss it in our next team meeting, I think it's a good idea - but I don't want to hold up the PR for it. I will make the change here for now. 
```suggestion
The following code gets the robot’s vision service and then runs a color detector vision model on an image from the robot's camera `"camera_1"`:
```
